Set up notifications once upon import and implement notification IDs

m-col opened the pull request:
A bool flag is used to track whether notifications are possible (i.e. if
gi cannot be imported), and send_notification will silently do nothing
if that are not. Notification IDs can be passed as a kwarg (id=4), which
the notification service can use to replace existing notifications
with the same ID.

ramnes commented on libqtile/utils.py:
```diff
@@ -25,10 +25,22 @@
+    if Notify.get_server_info()[0]:
+        _can_notify = True
+except Exception:
+    pass
```
Let's log something here?

ramnes commented:
Although it didn't look great, I think that the previous setup implementation was a great way to ensure we would always send a notification if possible. Why changing this?

m-col commented:
My motive was to reduce some function calls. Do you mean that the current implementation deals with the edge case where somebody installs `python-gobject` during qtile runtime and then it will work? Because why would we support that here and not in, for example, all the widgets that have external dependencies?

tych0 commented:
For what it's worth, I think it's totally fine to make people restart when they've installed new dependencies. I see little value and lots of code in doing it dynamically.

On Sun, Sep 13, 2020, at 10:01 AM, m-col wrote:
> 
> 
> My motive was to reduce some function calls. Do you mean that the 
> current implementation deals with the edge case where somebody installs 
> `python-gobject` during qtile runtime and then it will work? Because 
> why would we support that here and not in, for example, all the widgets 
> that have external dependencies?
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/qtile/qtile/pull/1901#issuecomment-691689854>, or 
> unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AAF7VV3DTH4ODP655ZEMF2LSFTUGPANCNFSM4RDIBA4Q>.
>

ramnes commented:
> Do you mean that the current implementation deals with the edge case where somebody installs `python-gobject` during qtile runtime and then it will work?

No, what worries me is what happens if the notification server is not up and running at the time of import, but shortly afterwards.

m-col commented:
Oh, I had not considered that. Good point.

I've updated the commit so that availability of a service is tested when sending a notification, so this should not be an issue. The imports can still be done ahead of time. I also added the logging call as you suggested.

tcld commented on libqtile/utils.py:
```diff
@@ -219,22 +230,19 @@ def class_proxy(*args, **kwargs):
+        notifier.set_timeout(timeout)
+        if urgent:
+            notifier.set_urgency(Notify.Urgency.CRITICAL)
+        notifier.set_timeout(timeout)
```
Why are you setting the timeout twice (four lines above it is also set)? If there is a reason, a comment would be important.

m-col replied:
Nope, that is a total typo! Thanks for spotting it.

tcld commented on libqtile/utils.py:
```diff
@@ -219,22 +230,19 @@ def class_proxy(*args, **kwargs):
+            notifier.set_urgency(Notify.Urgency.CRITICAL)
+        notifier.set_timeout(timeout)
+        if id is None:
+            id = randint(10, 1000)
```
I'm probably overthinking this (especially since we didn't have an id before^^), but maybe a uuid or a timestamp would be nicer here?

m-col replied:
This is per the spec (see https://developer.gnome.org/notification-spec/). The Replaces ID is a UINT32.

tcld replied:
You always have those links at the ready.^^ Maybe add it in a comment?

m-col replied:
Sure, and I've explained how this can be used :)

tcld replied:
That's great stuff, thanks. :)

tych0 commented on libqtile/utils.py:
```diff
@@ -25,10 +25,21 @@
+    from gi.repository import Notify  # type: ignore
+    Notify.init("Qtile")
+    _can_notify = True
+except Exception as e:
```
Can we use ImportError here if we only expect import errors?

m-col replied:
yep done.

tych0 commented:
lgtm, thanks!
